File: DebugIntro/C/02_CombiningBigAndSmall.c
#include <stdio.h>
#include <stdlib.h>
#include <float.h>
#include <time.h>

#define size 10
#define eg_float float

void fill_array(eg_float* array, int array_size);

double sum(eg_float* array, int array_sz);

float sum_as_float(eg_float* array, int array_sz);

eg_float sum_onto_value(eg_float* array, int array_sz, eg_float init_value);

int main(int argc, char** argv){

  int i = 0;
  eg_float array[size];
  double sum_d;
  float sum_f;
  eg_float sum_onto, init = 2.34642e5;

  fill_array(array, size);

  printf("Array of %d random values, printed to 3dp:\n", size);

  for(i=0; i<size; i++){
    printf("%6.3f\t", array[i]);
  }
  printf("\n");

  sum_d = sum(array, size);
  sum_f = sum_as_float(array, size);

  printf("\n");

  printf("Summed as floats (32bit) %12.10f and as double (64bit) %12.10f\n", sum_d, sum_f);
  printf("Difference is: %10.6e\n", sum_d-sum_f);

  sum_onto = sum_onto_value(array, size, init);

  printf("\n");


  printf("Starting with %10.6e and adding on the array values gives: %10.6e\n", init, sum_onto);
  printf("This gives an error of %10.6e in the sum\n",
     (sum_onto-init)-sum_f);
  /* If you change the define to double type, change the line above to use sum_d*/

  return 0;
}

double sum(eg_float* array, int array_sz){

  int i;
  double sum = 0.0;
  for(i=0; i< array_sz; i++){
    sum = sum + array[i];
  }
  return sum;
}

float sum_as_float(eg_float* array, int array_sz){

  int i;
  float sum = 0.0;
  for(i=0; i< array_sz; i++){
    sum = sum + (float)array[i];
  }
  return sum;
}

eg_float sum_onto_value(eg_float* array, int array_sz, eg_float init_value){
  /* Sums the given values along with the provided init value*/

  int i;
  eg_float sum = init_value;
  for(i=0; i< array_sz; i++){
    sum = sum + array[i];
  }
  return sum;
}

void fill_array(eg_float* array, int array_size){

  int i;

#  ifndef DEBUG
  srand(time(NULL)); /*Init RNG with current time*/
#  else
  srand(1287);/*Use a fixed seed */
#  endif

  /*Fill array with random values.
  For simplicity use rand / 100 */

  for(i=0; i< array_size; i++){
    array[i] = (rand()%10000)/3000.0;
  }
}

File: DebugIntro/Python/02_CombiningBigAndSmall.py
import random

def fill_array(array, use_fixed=False):

  if use_fixed:
    random.seed(1287)
  else:
    random.seed()

  for i in range(0, len(array)):
    array[i] = random.random()* 333.0

def sum_with_loop(array):

  total = 0.0
  for i in range(0, len(array)):
    total += array[i]
  return total

def sum_onto_value(array, init):

  total = init
  for i in range(0, len(array)):
    total += array[i]
  return total

if __name__ == "__main__":

  _sz = 10
  a = 3.0
  array = list(range(0, _sz))
  fill_array(array)

  print("Array of {0} random values, printed to 3dp:".format(_sz));

  print('\t'.join(["{0:6.3f}".format(x) for x in array]))

  sum_f = sum_with_loop(array)
  print("\nSum is {0:12.10f}".format(sum_f))

  init = 2.34642e5
  sum_onto = sum_onto_value(array, init)

  print("\nStarting with {0:12.10f} and adding on the array values gives {1:12.10f}".format(init, sum_onto))

  print("This gives an error of {0:12.10e} in the sum".format((sum_onto-init)-sum_f))


